src/authorization/WebAclReader.ts

```diff
@@ -56,17 +57,43 @@ export class WebAclReader extends PermissionReader {
    * Will throw an error if this is not the case.
    * @param input - Relevant data needed to check if access can be granted.
    */
-  public async handle({ identifier, credentials }: PermissionReaderInput):
+  public async handle({ identifier, credentials, modes }: PermissionReaderInput):
   Promise<PermissionSet> {
     // Determine the required access modes
     this.logger.debug(`Retrieving permissions of ${credentials.agent?.webId} for ${identifier.path}`);
 
     const isAcl = this.aclStrategy.isAuxiliaryIdentifier(identifier);
     const mainIdentifier = isAcl ? this.aclStrategy.getSubjectIdentifier(identifier) : identifier;
+    const checkParent = modes.has(AccessMode.create) || modes.has(AccessMode.delete);
 
-    // Determine the full authorization for the agent granted by the applicable ACL
-    const acl = await this.getAclRecursive(mainIdentifier);
-    return this.createPermissions(credentials, acl, isAcl);
+    // Determine the full authorization for the agent granted by the applicable ACL.
+    // Note that we don't filter on input modes as all results are needed for the WAC-Allow header.
+    const acl = await this.getAcl(mainIdentifier, checkParent);
+    const result = await this.createPermissions(credentials, acl.target, isAcl);
+
+    if (checkParent) {
+      this.logger.debug(`Verifying parent permissions due "create/delete" permission requirements`);
+      const parentPermissions = acl.target === acl.parent ?
```

Comment:
Oh that was a surprise to me; how can that happen again?

Reply:
Wait, it's `POST`, isn't it? I think we might need to rename.
```suggestion
      const containerPermissions = acl.target === acl.parent ?
```

Reply:
Note that this is comparing N3 stores, not identifiers. This is just a minor optimization in case they both use the same set of acl triples, which can occur if their permissions are in the same authorization in the same acl file (a block that has both an `acl:target` and an `acl:default` triple).